contrib/golang/upstreams/http/tcp/source/upstream_request.cc

```diff
@@ -0,0 +1,479 @@
+      throw EnvoyException(fmt::format("tcp upstream : load library failed: {} {}", c.library_id(), c.library_path()));
+    };
+
+    FilterConfigSharedPtr conf = std::make_shared<FilterConfig>(c, dynamic_lib_);
+    conf->newGoPluginConfig();
+    config_ = conf;
+    plugin_name_ = c.plugin_name();
+}
+
+void TcpConnPool::onPoolReady(Envoy::Tcp::ConnectionPool::ConnectionDataPtr&& conn_data,
+                              Upstream::HostDescriptionConstSharedPtr host) {
+  upstream_handle_ = nullptr;
+  Network::Connection& latched_conn = conn_data->connection();
+  auto upstream =
+      std::make_unique<TcpUpstream>(&callbacks_->upstreamToDownstream(), std::move(conn_data), dynamic_lib_, config_);
+
+  ENVOY_LOG(debug, "tcp upstream get host info: {}", host->cluster().name());
+
+  callbacks_->onPoolReady(std::move(upstream), host, latched_conn.connectionInfoProvider(),latched_conn.streamInfo(), {});       
+}
+
+TcpUpstream::TcpUpstream(Router::UpstreamToDownstream* upstream_request,
+                         Envoy::Tcp::ConnectionPool::ConnectionDataPtr&& upstream, Dso::TcpUpstreamDsoPtr dynamic_lib,
+                         FilterConfigSharedPtr config)
+    :encoding_state_(new EncodingProcessorState(*this)), decoding_state_(new DecodingProcessorState(*this)),
+     route_entry_(upstream_request->route().routeEntry()), upstream_request_(upstream_request), upstream_conn_data_(std::move(upstream)), dynamic_lib_(dynamic_lib), config_(config) {
+  
+  configId = config->getConfigId();
+  plugin_name.data = config->pluginName().data();
+  plugin_name.len = config->pluginName().length();
+
+  upstream_conn_data_->connection().enableHalfClose(true);
+  upstream_conn_data_->addUpstreamCallbacks(*this);
+}
+
+TcpUpstream::~TcpUpstream() {
+  ENVOY_LOG(debug, "tcp upstream on destroy");
+
+  // initRequest haven't be called yet, which mean haven't called into Go.
+  if (configId == 0) {
+    return;
+  }              
+
+  auto reason = (decoding_state_->isProcessingInGo() || encoding_state_->isProcessingInGo())
+                ? DestroyReason::Terminate
+                : DestroyReason::Normal;
+  dynamic_lib_->envoyGoOnTcpUpstreamDestroy(this, int(reason));
+}
+
+void TcpUpstream::initRequest() {
+  if (configId == 0) {
+    configId = config_->getConfigId();
+  }
+};
+
+void TcpUpstream::initResponse() {
+  if (resp_headers_ == nullptr) {
+    auto headers{Envoy::Http::createHeaderMap<Envoy::Http::ResponseHeaderMapImpl>({})};
+    resp_headers_ = std::move(headers);
+  }
+}
+
+Envoy::Http::Status TcpUpstream::encodeHeaders(const Envoy::Http::RequestHeaderMap& headers,
+                                               bool end_stream) {
+  ENVOY_LOG(debug, "tcp upstream encodeHeaders, header size: {}, end_stream: {}", headers.size(), end_stream);
+
+  initRequest();
+
+  ProcessorState* state = encoding_state_;
+  state->headers = &headers;
+  Buffer::OwnedImpl buf;
+  Buffer::Instance& buffer = state->doDataList.push(buf);
```

Comment:
do not need to push to `doDataList`, since there is only one data buffer.

Reply:
done